lib/Transforms/IPO/StructAnalysisTransformAnalyzer.cpp

```diff
@@ -0,0 +1,865 @@
+                  dbgs() << "Initializing StructSplitAnalyzer of size: "
+                         << NumElements << "\n");
+  // Calculate every CP relations and save it to an array to avoid repeated
+  // calculation
+  CloseProximityRelations.resize(NumElements);
+  for (unsigned i = 0; i < NumElements; i++) {
+    CloseProximityRelations[i].resize(NumElements);
+    for (unsigned j = 0; j < NumElements; j++) {
+      CloseProximityRelations[i][j] = calculateCloseProximity(i, j);
+    }
+  }
+  for (unsigned i = 0; i < NumElements; i++) {
+    DEBUG(dbgs() << "F" << i + 1 << "\t");
+    for (unsigned j = 0; j < NumElements; j++)
+      DEBUG(dbgs() << DEBUG_PRINT_COUNT(CloseProximityRelations[i][j]) << "\t");
+    DEBUG(dbgs() << "\n");
+  }
+
+  // Debug print all parameters used in this suggestion
+  DEBUG(dbgs() << "Parameters used in struct spliting: \n");
+  DEBUG(dbgs() << "COLD_RATIO: " << Params.ColdRatio << "\n");
+  DEBUG(dbgs() << "DISTANCE_THRESHOLD: " << Params.DistanceThreshold << "\n");
+  DEBUG(dbgs() << "MAX_SIZE: " << Params.MaxSize << "\n");
+  DEBUG(dbgs() << "SIZE_PENALTY: " << Params.SizePenalty << "\n");
+  DEBUG(dbgs() << "CPG for struct split recommendation\n");
+
+  DEBUG_WITH_TYPE(DEBUG_TYPE_SPLIT, dbgs() << "Finish constructor\n");
+}
+
+// Function to calculate CP value based on CloseProximityBuilder passed in
+// Return a floating point value equals to a portion of the count part of
+// the CP relations. The portion (SizePenalty) is used to penalize two fields
+// with different sizes. (Derived from the paper). Special cases are
+// when distance of the pair is larger than a threshold (DistanceThreshold),
+// the pair is no longer considered "close" and same applies to when two
+// fields are too large (larger than MaxSize).
+double StructSplitAnalyzer::calculateCloseProximity(FieldNumType Field1,
+                                                    FieldNumType Field2) const {
+  auto *Pair = CPBuilder->getCloseProximityPair(Field1, Field2);
+  if (Pair->second > Params.DistanceThreshold)
+    return 0;
+  if (FieldSizes[Field1] + FieldSizes[Field2] > Params.MaxSize)
+    return 0;
+  auto Ratio = 0;
+  if (FieldSizes[Field1] > FieldSizes[Field2])
+    Ratio = FieldSizes[Field1] / FieldSizes[Field2];
+  else
+    Ratio = FieldSizes[Field2] / FieldSizes[Field1];
+  Ratio *= Params.SizePenalty;
+  return Pair->first / Ratio;
+}
+
+// Function to find maximum CP value of all pairs in remaining fields
+// Return a std::pair type of the two fields
+FieldPairType StructSplitAnalyzer::findMaxRemainCP() const {
+  double MaxCP = 0;
+  auto it = FieldsToTransform.begin();
+  auto Field1 = *(it++);
+  assert(it != FieldsToTransform.end());
+  auto Field2 = *it;
+  DEBUG_WITH_TYPE(DEBUG_TYPE_SPLIT,
+                  dbgs() << "Try to find a pair with maxmium CP\n");
+  for (auto &F1 : FieldsToTransform)
+    for (auto &F2 : FieldsToTransform)
+      if (F1 != F2) {
+        if (CloseProximityRelations[F1][F2] > MaxCP) {
```

Comment:
Yes. I'll change it to F1 < F2 to make sure only one update happens to MaxCP and the pair is always has field with smaller number comes first. Thanks!